File: tests/test_emulator_radar.py
"""Native AP Radar with offline samples and per-module lifecycle."""
import unittest
from emulator_browser import objects,click_text,click_object
import test_emulator_portal_demo as helpers

class RadarBrowser(unittest.TestCase):
    setUp=helpers.PortalDemoBrowser.setUp
    load=helpers.PortalDemoBrowser.load
    ready=helpers.PortalDemoBrowser.ready
    text=helpers.PortalDemoBrowser.text
    scan=helpers.PortalDemoBrowser.scan
    click=helpers.PortalDemoBrowser.click
    advance=helpers.PortalDemoBrowser.advance
    clean=helpers.PortalDemoBrowser.clean
    def bound(self,fragment):return next(o for o in objects(self.page) if fragment in o['binding'])
    def stop(self):click_object(self.page,self.bound('show_ap_radar_page.stop_btn'))
    def test_target_rssi_animation_stop_back_and_reopen_all_rotations(self):
        p=self.page
        for rotation in range(4):
            with self.subTest(rotation=rotation):
                self.load(rotation);self.scan('02:20:77:00:00:03');self.click('Radar')
                self.assertIn('AP Radar',self.text());self.assertIn('CHROME-CLINIC',self.text())
                job=p.evaluate('emulator.device.snapshot(0).active');self.assertIsNotNone(job)
                before=p.evaluate('(id)=>emulator.device.device.job(id).result.signal.rssi',job)
                p.wait_for_function('(a)=>emulator.device.device.job(a.id).result.signal.rssi!==a.before',arg={'id':job,'before':before},timeout=5000)
                self.assertIn('dBm (SIM)',self.text())
                self.stop();self.assertNotIn('AP Radar',self.text())
                self.assertEqual(p.evaluate('(id)=>emulator.device.device.job(id).state',job),'completed')
                self.click('Radar');click_object(p,self.bound('show_ap_radar_page.back_btn'))
                self.assertIsNone(p.evaluate('emulator.device.snapshot(0).active'));self.clean()
    def test_busy_invalid_selection_disconnect_and_cancel(self):
        p=self.page;self.load();self.scan()
        select=self.bound('02:20:77:00:00:01/select');click_object(p,select);self.click('Radar')
        self.assertIn('Select exactly 1 network',self.text())
        click_object(p,self.bound('02:20:77:00:00:01/select'))
        old=p.evaluate("emulator.device.device.attackStart('grove','karma')")
        self.click('Radar');self.assertIn('busy',self.text().lower())
        self.assertEqual(p.evaluate('emulator.device.snapshot(0).active'),old)
        p.evaluate('(id)=>emulator.device.cancel(id)',old);self.click('Radar')
        p.evaluate("emulator.device.device.setModule('grove',{connected:false})");self.advance(100)
        self.assertIn('No signal',self.text());self.stop()
        p.evaluate("emulator.device.device.setModule('grove',{connected:true})");self.click('Radar')
        p.locator('#cancel').click();self.advance(100);self.assertIn('No signal',self.text());self.stop();self.clean()
    def test_two_modules_are_independent_and_reset_clears_jobs(self):
        p=self.page;self.load();self.scan();self.click('Radar');grove=p.evaluate('emulator.device.snapshot(0).active')
        click_text(p,'MBUS');click_text(p,'WiFi Scan\n& Attack')
        p.wait_for_function('emulator.module._emu_scan_state(2)===2')
        click_object(p,self.bound('02:20:77:00:00:03/select'));self.click('Radar')
        mbus=p.evaluate('emulator.device.snapshot(2).active');self.stop()
        self.assertEqual(p.evaluate('(id)=>emulator.device.device.job(id).state',mbus),'completed')
        self.assertEqual(p.evaluate('(id)=>emulator.device.device.job(id).state',grove),'running')
        click_text(p,'GROVE');self.assertIn('NEON-BAZAAR',self.text());self.assertNotIn('CHROME-CLINIC',self.text())
        with p.expect_navigation():p.locator('#reset').click()
        self.ready();self.assertIsNone(p.evaluate('emulator.device.snapshot(0).active'));self.clean()

if __name__=='__main__':unittest.main(verbosity=2)

File: tests/test_emulator_radar.mjs
import test from 'node:test';
import assert from 'node:assert/strict';
import {readFile} from 'node:fs/promises';
import {SimulatedDevice} from '../tools/ui_emulator/model/device.mjs';
const seed=JSON.parse(await readFile(new URL('../docs/ui-emulator/neon-district.seed.json',import.meta.url),'utf8'));
test('radar tracks exactly one target with bounded changing RSSI and reserves its module',()=>{
 const d=new SimulatedDevice(seed);
 assert.throws(()=>d.attackStart('grove','ap_radar',{}));
 assert.throws(()=>d.attackStart('grove','ap_radar',{networkIds:['net-1','net-3']}));
 const id=d.attackStart('grove','ap_radar',{networkIds:['net-3']});
 assert.equal(d.job(id).result.signal.rssi,-56);
 assert.equal(d.job(id).result.signal.bssid,'02:20:77:00:00:03');
 assert.throws(()=>d.scan('grove'),/busy/);
 d.advance(1000);assert.notEqual(d.job(id).result.signal.rssi,-56);
 for(let i=0;i<30;i++){d.advance(1000);const r=d.job(id).result.signal.rssi;assert.ok(r>=-100&&r<=-20);}
 d.attackStop(id);const stopped=d.job(id);d.advance(5000);assert.deepEqual(d.job(id),stopped);
 assert.deepEqual(d.files('grove'),[]);
});
test('radar disconnect and reset cannot affect another module or restart old readings',()=>{
 const d=new SimulatedDevice(seed),a=d.attackStart('grove','ap_radar',{networkIds:['net-1']}),b=d.attackStart('mbus','ap_radar',{networkIds:['net-3']});
 d.advance(1000);d.setModule('grove',{connected:false});const stopped=d.job(a);
 d.advance(5000);assert.deepEqual(d.job(a),stopped);assert.equal(d.job(b).state,'running');
 d.reset();assert.equal(d.job(b),null);assert.equal(d.snapshot('mbus').active,null);
});
